File: src/markov/WordGraph.java
package markov;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.util.Scanner;

public class WordGraph {
    protected WeightedGraph<String> graph;

    protected String lastWord;

    public WordGraph() {
        graph = new WeightedAdjacencyListGraph<>();
        lastWord = null;
    }

    public void addWord(String newWord) {
        //TODO: add a word to the graph instance variable
        var trimmedWord = newWord.trim();
        if (!trimmedWord.equals("")) {
            graph.add(trimmedWord);
            if (lastWord != null) {
                graph.addEdge(lastWord, trimmedWord);
                graph.setWeight(lastWord, trimmedWord, graph.getWeight(lastWord, trimmedWord) + 1);
            }
            lastWord = trimmedWord;
        }
    }
    
    public void addWordsFromString(String str) {
        for (String word : str.split(" "))
            addWord(word);
    }


    public void processFile(String filename) {
        try {
            for(String line: new BufferedReader(new FileReader((filename))).lines().toList()) {
                addWordsFromString(line);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public WeightedGraph<String> getGraph() {
        return graph;
    }

}

File: src/markov_kt/WordGraph.kt
package markov_kt

import markov.WeightedAdjacencyListGraph
import markov.WeightedGraph
import java.io.BufferedReader
import java.io.FileReader

internal class WordGraph {
    var graph: WeightedGraph<String>
        protected set
    protected var lastWord: String?

    init {
        graph = WeightedAdjacencyListGraph()
        lastWord = null
    }

    fun addWord(newWord: String) {
        //TODO: add a word to the graph instance variable
        val trimmedWord = newWord.trim { it <= ' ' }
        if (trimmedWord != "") {
            graph.add(trimmedWord)
            if (lastWord != null) {
                graph.addEdge(lastWord, trimmedWord)
                graph.setWeight(lastWord, trimmedWord, graph.getWeight(lastWord, trimmedWord) + 1)
            }
            lastWord = trimmedWord
        }
    }

    fun addWordsFromString(str: String) {
        for (word in str.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()) addWord(word)
    }

    fun processFile(filename: String?) {
        try {
            for (line in BufferedReader(FileReader(filename)).lines().toList()) {
                addWordsFromString(line)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}